README.md
## signature-mobile-wechat

signature-mobile-wechat support H5 and wechat miniprogram

This is a rewrite of original signature-mobile-wechat, starting and borrowing from [this fork](https://github.com/szimek/signature_pad).

## Installation

```bash
# with npm
npm install signature-mobile-wechat -S
# with yarn
yarn add signature-mobile-wechat -S
```

## Usage

### 1-导入包
```typescript
import { SignatureMobile, SignatureMobileWechat } from 'signature-mobile-wechat';
```
### 2-在微信小程序中使用
```
// 1-创建初始化
wx.createSelectorQuery()
  .select(`#canvas`)
  .fields({
    node: true,
    size: true,
  })
  .exec((res: any[]) => {
    const canvas: WechatMiniprogram.Canvas = res[0].node;
    const signatureMobileWechat = new SignatureMobileWechat(canvas, {
      minWidth: 2,
      maxWidth: 5,
      penColor: 'rgb(0, 0, 0)',
    });
  });
// 2-手势注册注入
handleTouchStart: function (e) {
  // console.log(e)
  this.data.signatureMobileWechat && this.data.signatureMobileWechat.handleTouchStart(e)
},
handleTouchMove: function (e) {
  // console.log(e)
  this.data.signatureMobileWechat && this.data.signatureMobileWechat.handleTouchMove(e)
},
handleTouchEnd: function (e) {
  // console.log(e)
  this.data.signatureMobileWechat && this.data.signatureMobileWechat.handleTouchEnd(e)
},
```
### 2-正常网页使用
```
const canvas = document.querySelector('canvas');
const signatureMobile = new SignatureMobile(canvas, {
  minWidth: 2,
  maxWidth: 5,
  penColor: 'rgb(0, 0, 0)',
});
```

### Options
<dl>
<dt>dotSize</dt>
<dd>(float or function) Radius of a single dot.</dd>
<dt>minWidth</dt>
<dd>(float) Minimum width of a line. Defaults to <code>0.5</code>.</dd>
<dt>maxWidth</dt>
<dd>(float) Maximum width of a line. Defaults to <code>2.5</code>.</dd>
<dt>throttle</dt>
<dd>(integer) Draw the next point at most once per every <code>x</code> milliseconds. Set it to <code>0</code> to turn off throttling. Defaults to <code>16</code>.</dd>
<dt>minDistance</dt>
<dd>(integer) Add the next point only if the previous one is farther than <code>x</code> pixels. Defaults to <code>5</code>.
<dt>backgroundColor</dt>
<dd>(string) Color used to clear the background. Can be any color format accepted by <code>context.fillStyle</code>. Defaults to <code>"rgba(0,0,0,0)"</code> (transparent black). Use a non-transparent color e.g. <code>"rgb(255,255,255)"</code> (opaque white) if you'd like to save signatures as JPEG images.</dd>
<dt>penColor</dt>
<dd>(string) Color used to draw the lines. Can be any color format accepted by <code>context.fillStyle</code>. Defaults to <code>"black"</code>.</dd>
<dt>velocityFilterWeight</dt>
<dd>(float) Weight used to modify new velocity based on the previous velocity. Defaults to <code>0.7</code>.</dd>
<dt>onBegin</dt>
<dd>(function) Callback when stroke begin.</dd>
<dt>onEnd</dt>
<dd>(function) Callback when stroke end.</dd>
</dl>

You can set options during initialization:
```javascript
var signatureMobile = new SignatureMobile(canvas, {
    minWidth: 5,
    maxWidth: 10,
    penColor: "rgb(66, 133, 244)"
});
```
or during runtime:
```javascript
var signatureMobile = new SignatureMobile(canvas);
signatureMobile.minWidth = 5;
signatureMobile.maxWidth = 10;
signatureMobile.penColor = "rgb(66, 133, 244)";
```

CHANGELOG.md
## Changelog

### 0.0.6

`2021-07-26`

- Feat
 - 添加minWidth、maxWidth

### 0.0.4 and 0.0.5

`2021-05-11`

- Feat
 - ts

### 0.0.3

`2021-02-19`

- Feat
 - 添加dot-width
### 0.0.2

`2020-12-28`

- Fix
 - 修复update信息(sendMesg)回掉可能重复触发问题

### 0.0.1

`2020-12-25`

- init


rollup.config.js
const typescript = require('rollup-plugin-typescript2');
const { terser } = require('rollup-plugin-terser');
const pkg = require('./package.json');

const banner =
  '/*!\n' +
  ` * Signature Pad v${pkg.version} | ${pkg.homepage}\n` +
  ` * (c) ${new Date().getFullYear()} ${
    pkg.author.name
  } | Released under the MIT license\n` +
  ' */\n';

export default [
  {
    // UMD unminified
    input: 'src/index.ts',
    plugins: [typescript({
      target: 'ES3',
      useTsconfigDeclarationDir: true,
    })],
    output: {
      // dir: 'dist',
      file: 'dist/signature-mobile-wechat.umd.js',
      format: 'umd',
      name: 'Signature',
      sourcemap: true,
      banner,
    },
  },
  {
    // UMD unminified
    input: 'src/index.ts',
    plugins: [typescript({
      target: 'ES3',
      useTsconfigDeclarationDir: true,
    })],
    output: {
      // dir: 'dist',
      file: 'dist/signature-mobile-wechat.umd.min.js',
      format: 'umd',
      name: 'Signature',
      sourcemap: true,
      plugins: [terser()],
      banner,
    },
  },
  {
    // ES module unminified
    input: 'src/index.ts',
    plugins: [typescript({
      target: 'ES2015',
      useTsconfigDeclarationDir: true,
    })],
    output: {
      // dir: 'dist',
      file: 'dist/signature-mobile-wechat.js',
      format: 'es',
      sourcemap: true,
      banner,
    },
  },
  {
    // ES module minified
    input: 'src/index.ts',
    plugins: [typescript({
      target: 'ES2015',
      useTsconfigDeclarationDir: true,
    })],
    output: {
      // dir: 'dist',
      file: 'dist/signature-mobile-wechat.min.js',
      format: 'es',
      sourcemap: true,
      plugins: [terser()],
      banner,
    },
  }
];


src/signature-mobile.ts
/**
 * The main idea and some parts of the code (e.g. drawing variable width Bézier curve) are taken from:
 * http://corner.squareup.com/2012/07/smoother-signatures.html
 *
 * Implementation of interpolation using cubic Bézier curves is taken from:
 * http://www.benknowscode.com/2012/09/path-interpolation-using-cubic-bezier_9742.html
 *
 * Algorithm for approximated length of a Bézier curve is taken from:
 * http://www.lemoda.net/maths/bezier-length/index.html
 */

// import { Bezier } from './bezier';
// import { BasicPoint, Point } from './point';
import SignaturePadBase, { Options } from './signature-base';

export default class SignatureMobile extends SignaturePadBase {
  public onBegin?: (event: MouseEvent | Touch) => void;
  public onEnd?: (event: MouseEvent | Touch) => void;
  public canvas: HTMLCanvasElement;

  // Private stuff
  /* tslint:disable: variable-name */
  private _mouseButtonDown: boolean;

  constructor(canvas: HTMLCanvasElement, options: Options = {}, sendMsg?: Function) {
    super(canvas, options, sendMsg);
    this.onBegin = options.onBegin;
    this.onEnd = options.onEnd;

    // Enable mouse and touch event handlers
    this.on();
  }

  public fromDataURL(
    dataUrl: string,
    options: { ratio?: number; width?: number; height?: number } = {},
    callback?: (error?: string | Event) => void,
  ): void {
    const image = new Image();
    const ratio = options.ratio || window.devicePixelRatio || 1;
    const width = options.width || this.canvas.width / ratio;
    const height = options.height || this.canvas.height / ratio;

    this.reset();

    image.onload = (): void => {
      this.ctx.drawImage(image, 0, 0, width, height);
      if (callback) {
        callback();
      }
    };
    image.onerror = (error): void => {
      if (callback) {
        callback(error);
      }
    };
    image.src = dataUrl;

    this._isEmpty = false;
  }

  // public toDataURL(type = 'image/png', encoderOptions?: number): string {
  //   switch (type) {
  //     case 'image/svg+xml':
  //       return this._toSVG();
  //     default:
  //       return this.canvas.toDataURL(type, encoderOptions);
  //   }
  // }

  public on(): void {
    // Disable panning/zooming when touching canvas element
    this.canvas.style.touchAction = 'none';
    this.canvas.style.msTouchAction = 'none';

    if (window.PointerEvent) {
      this._handlePointerEvents();
    } else {
      this._handleMouseEvents();

      if ('ontouchstart' in window) {
        this._handleTouchEvents();
      }
    }
  }

  public off(): void {
    // Enable panning/zooming when touching canvas element
    this.canvas.style.touchAction = 'auto';
    this.canvas.style.msTouchAction = 'auto';

    this.canvas.removeEventListener('pointerdown', this._handleMouseDown);
    this.canvas.removeEventListener('pointermove', this._handleMouseMove);
    document.removeEventListener('pointerup', this._handleMouseUp);

    this.canvas.removeEventListener('mousedown', this._handleMouseDown);
    this.canvas.removeEventListener('mousemove', this._handleMouseMove);
    document.removeEventListener('mouseup', this._handleMouseUp);

    this.canvas.removeEventListener('touchstart', this._handleTouchStart);
    this.canvas.removeEventListener('touchmove', this._handleTouchMove);
    this.canvas.removeEventListener('touchend', this._handleTouchEnd);
  }

  public isEmpty(): boolean {
    return this._isEmpty;
  }

  // public fromData(pointGroups: PointGroup[]): void {
  //   this.clear();

  //   this._fromData(
  //     pointGroups,
  //     ({ color, curve }) => this._drawCurve({ color, curve }),
  //     ({ color, point }) => this._drawDot({ color, point }),
  //   );

  //   this.data = pointGroups;
  // }

  // public toData(): PointGroup[] {
  //   return this.data;
  // }

  // Event handlers
  private _handleMouseDown = (event: MouseEvent): void => {
    if (event.which === 1) {
      this._mouseButtonDown = true;
      this.strokeBegin(event);
    }
  };

  private _handleMouseMove = (event: MouseEvent): void => {
    if (this._mouseButtonDown) {
      this.strokeMoveUpdate(event);
    }
  };

  private _handleMouseUp = (event: MouseEvent): void => {
    if (event.which === 1 && this._mouseButtonDown) {
      this._mouseButtonDown = false;
      this.strokeEnd(event);
    }
  };

  private _handleTouchStart = (event: TouchEvent): void => {
    // Prevent scrolling.
    event.preventDefault();

    if (event.targetTouches.length === 1) {
      const touch = event.changedTouches[0];
      this.strokeBegin(touch);
    }
  };

  private _handleTouchMove = (event: TouchEvent): void => {
    // Prevent scrolling.
    event.preventDefault();

    const touch = event.targetTouches[0];
    this.strokeMoveUpdate(touch);
  };

  private _handleTouchEnd = (event: TouchEvent): void => {
    const wasCanvasTouched = event.target === this.canvas;
    if (wasCanvasTouched) {
      event.preventDefault();

      const touch = event.changedTouches[0];
      this.strokeEnd(touch);
    }
  };

  private _handlePointerEvents(): void {
    this._mouseButtonDown = false;

    this.canvas.addEventListener('pointerdown', this._handleMouseDown);
    this.canvas.addEventListener('pointermove', this._handleMouseMove);
    document.addEventListener('pointerup', this._handleMouseUp);
  }

  private _handleMouseEvents(): void {
    this._mouseButtonDown = false;

    this.canvas.addEventListener('mousedown', this._handleMouseDown);
    this.canvas.addEventListener('mousemove', this._handleMouseMove);
    document.addEventListener('mouseup', this._handleMouseUp);
  }

  private _handleTouchEvents(): void {
    this.canvas.addEventListener('touchstart', this._handleTouchStart);
    this.canvas.addEventListener('touchmove', this._handleTouchMove);
    this.canvas.addEventListener('touchend', this._handleTouchEnd);
  }

  // private _fromData(
  //   pointGroups: PointGroup[],
  //   drawCurve: SignatureMobile['_drawCurve'],
  //   drawDot: SignatureMobile['_drawDot'],
  // ): void {
  //   for (const group of pointGroups) {
  //     const { color, points } = group;

  //     if (points.length > 1) {
  //       for (let j = 0; j < points.length; j += 1) {
  //         const basicPoint = points[j];
  //         const point = new Point(basicPoint.x, basicPoint.y, basicPoint.time);

  //         // All points in the group have the same color, so it's enough to set
  //         // penColor just at the beginning.
  //         this.penColor = color;

  //         if (j === 0) {
  //           this.reset();
  //         }

  //         const curve = this._addPoint(point);

  //         if (curve) {
  //           drawCurve({ color, curve });
  //         }
  //       }
  //     } else {
  //       this.reset();

  //       drawDot({
  //         color,
  //         point: points[0],
  //       });
  //     }
  //   }
  // }

  // private _toSVG(): string {
  //   const pointGroups = this.data;
  //   const ratio = Math.max(window.devicePixelRatio || 1, 1);
  //   const minX = 0;
  //   const minY = 0;
  //   const maxX = this.canvas.width / ratio;
  //   const maxY = this.canvas.height / ratio;
  //   const svg = document.createElementNS('http://www.w3.org/2000/svg', 'svg');

  //   svg.setAttribute('width', this.canvas.width.toString());
  //   svg.setAttribute('height', this.canvas.height.toString());

  //   this._fromData(
  //     pointGroups,

  //     ({ color, curve }: { color: string; curve: Bezier }) => {
  //       const path = document.createElement('path');

  //       // Need to check curve for NaN values, these pop up when drawing
  //       // lines on the canvas that are not continuous. E.g. Sharp corners
  //       // or stopping mid-stroke and than continuing without lifting mouse.
  //       /* eslint-disable no-restricted-globals */
  //       if (
  //         !isNaN(curve.control1.x) &&
  //         !isNaN(curve.control1.y) &&
  //         !isNaN(curve.control2.x) &&
  //         !isNaN(curve.control2.y)
  //       ) {
  //         const attr =
  //           `M ${curve.startPoint.x.toFixed(3)},${curve.startPoint.y.toFixed(
  //             3,
  //           )} ` +
  //           `C ${curve.control1.x.toFixed(3)},${curve.control1.y.toFixed(3)} ` +
  //           `${curve.control2.x.toFixed(3)},${curve.control2.y.toFixed(3)} ` +
  //           `${curve.endPoint.x.toFixed(3)},${curve.endPoint.y.toFixed(3)}`;
  //         path.setAttribute('d', attr);
  //         path.setAttribute('stroke-width', (curve.endWidth * 2.25).toFixed(3));
  //         path.setAttribute('stroke', color);
  //         path.setAttribute('fill', 'none');
  //         path.setAttribute('stroke-linecap', 'round');

  //         svg.appendChild(path);
  //       }
  //       /* eslint-enable no-restricted-globals */
  //     },

  //     ({ color, point }: { color: string; point: BasicPoint }) => {
  //       const circle = document.createElement('circle');
  //       const dotSize =
  //         typeof this.dotSize === 'function' ? this.dotSize() : this.dotSize;
  //       circle.setAttribute('r', dotSize.toString());
  //       circle.setAttribute('cx', point.x.toString());
  //       circle.setAttribute('cy', point.y.toString());
  //       circle.setAttribute('fill', color);

  //       svg.appendChild(circle);
  //     },
  //   );

  //   const prefix = 'data:image/svg+xml;base64,';
  //   const header =
  //     '<svg' +
  //     ' xmlns="http://www.w3.org/2000/svg"' +
  //     ' xmlns:xlink="http://www.w3.org/1999/xlink"' +
  //     ` viewBox="${minX} ${minY} ${maxX} ${maxY}"` +
  //     ` width="${maxX}"` +
  //     ` height="${maxY}"` +
  //     '>';
  //   let body = svg.innerHTML;

  //   // IE hack for missing innerHTML property on SVGElement
  //   if (body === undefined) {
  //     const dummy = document.createElement('dummy');
  //     const nodes = svg.childNodes;
  //     dummy.innerHTML = '';

  //     // tslint:disable-next-line: prefer-for-of
  //     for (let i = 0; i < nodes.length; i += 1) {
  //       dummy.appendChild(nodes[i].cloneNode(true));
  //     }

  //     body = dummy.innerHTML;
  //   }

  //   const footer = '</svg>';
  //   const data = header + body + footer;

  //   return prefix + btoa(data);
  // }
}


src/signature-mobile-wechat.ts
import SignaturePadBase, { Options, PointGroup } from './signature-base';
import { Point } from './point';

export default class SignatureMobileWechat extends SignaturePadBase {
  constructor(canvas: WechatMiniprogram.Canvas, options: Options = {}, sendMsg?: Function) {
    super(canvas, options, sendMsg);
  }
  public handleTouchStart = (event: WechatMiniprogram.Event): void => {
    if (event.touches.length === 1) {
      // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
      // @ts-ignore wechatMiniProgram declaration error
      const touch = event.changedTouches[0];
      this.strokeBegin(touch);
    }
  };

  public handleTouchMove = (event: WechatMiniprogram.Event): void => {
    // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
    // @ts-ignore wechatMiniProgram declaration error
    const touch = event.changedTouches[0];
    this.strokeMoveUpdate(touch);
  };

  public handleTouchEnd = (event: WechatMiniprogram.Event): void => {
    // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
    // @ts-ignore wechatMiniProgram declaration error
    const touch = event.changedTouches[0];
    this.strokeEnd(touch);
  };
}


src/signature-base.ts
/**
 * The main idea and some parts of the code (e.g. drawing variable width Bézier curve) are taken from:
 * http://corner.squareup.com/2012/07/smoother-signatures.html
 *
 * Implementation of interpolation using cubic Bézier curves is taken from:
 * http://www.benknowscode.com/2012/09/path-interpolation-using-cubic-bezier_9742.html
 *
 * Algorithm for approximated length of a Bézier curve is taken from:
 * http://www.lemoda.net/maths/bezier-length/index.html
 */

import { Bezier } from './bezier';
import { BasicPoint, Point } from './point';
import { throttle } from './throttle';

declare global {
  interface Window {
    PointerEvent: typeof PointerEvent;
  }
  interface CSSStyleDeclaration {
    msTouchAction: string | null;
  }
}

export interface Options {
  dotSize?: number | (() => number);
  minWidth?: number;
  maxWidth?: number;
  minDistance?: number;
  backgroundColor?: string;
  penColor?: string;
  throttle?: number;
  velocityFilterWeight?: number;
  onBegin?: (event: MouseEvent | Touch | WechatMiniprogram.Touch) => void;
  onEnd?: (event: MouseEvent | Touch | WechatMiniprogram.Touch) => void;
}

export interface PointGroup {
  color: string;
  minWidth?: number;
  points: BasicPoint[];
}

export default class SignaturePadBase {
  // Public stuff
  public dotSize: number | (() => number);
  /** 笔迹最小宽度 */
  public minWidth: number;
  /** 笔迹最大宽度 */
  public maxWidth: number;
  public minDistance: number;
  public backgroundColor: string;
  public penColor: string;
  public throttle: number;
  public velocityFilterWeight: number;

  /* tslint:disable: variable-name */
  public ctx: CanvasRenderingContext2D;
  public lastPoints: Point[]; // Stores up to 4 most recent points; used to generate a new curve
  public data: PointGroup[]; // Stores all points in groups (one group per line or dot)
  public lastVelocity: number;
  public lastWidth: number;
  /** 手写笔迹绘制 */
  protected strokeMoveUpdate: (
    event: MouseEvent | Touch | WechatMiniprogram.Touch,
  ) => void;

  /* tslint:enable: variable-name */
  protected _isEmpty: boolean;

  constructor(
    protected canvas: HTMLCanvasElement | WechatMiniprogram.Canvas,
    protected options: Options = {},
    protected sendMsg?: Function,
  ) {
    this.velocityFilterWeight = options.velocityFilterWeight || 0.7;
    this.minWidth = options.minWidth || 0.5;
    this.maxWidth = options.maxWidth || 2.5;
    this.throttle = ('throttle' in options ? options.throttle : 16) as number; // in milisecondss
    this.minDistance = ('minDistance' in options
      ? options.minDistance
      : 5) as number; // in pixels
    this.dotSize =
      options.dotSize ||
      function dotSize(this: SignaturePadBase): number {
        return (this.minWidth + this.maxWidth) / 2;
      };
    this.penColor = options.penColor || 'black';
    this.backgroundColor = options.backgroundColor || 'rgba(0,0,0,0)';

    this.strokeMoveUpdate = this.throttle
      ? throttle(SignaturePadBase.prototype._strokeUpdate, this.throttle)
      : SignaturePadBase.prototype._strokeUpdate;
    this.ctx = canvas.getContext('2d') as CanvasRenderingContext2D;

    this.clear();
  }

  public clear(): void {
    const { ctx: ctx, canvas } = this;

    // Clear canvas using background color
    ctx.fillStyle = this.backgroundColor;
    ctx.clearRect(0, 0, canvas.width, canvas.height);
    ctx.fillRect(0, 0, canvas.width, canvas.height);

    this.data = [];
    this.reset();
    this._isEmpty = true;
  }

  public isEmpty(): boolean {
    return this._isEmpty;
  }

  public fromData(pointGroups: PointGroup[]): void {
    this.clear();

    this._fromData(
      pointGroups,
      ({ color, curve }) => this._drawCurve({ color, curve }),
      ({ color, point }) => this._drawDot({ color, point }),
    );

    this.data = pointGroups;
  }

  public toData(): PointGroup[] {
    return this.data;
  }

  public toDataURL(type = 'image/png', encoderOptions: number): string {
    switch (type) {
      case 'image/svg+xml':
        return this._toSVG();
      default:
        return this.canvas.toDataURL(type, encoderOptions);
    }
  }

  // Private methods
  public strokeBegin(
    event: MouseEvent | Touch | WechatMiniprogram.Touch,
  ): void {
    const newPointGroup = {
      id: new Date().getTime(),
      color: this.penColor,
      maxWidth: this.maxWidth,
      minWidth: this.minWidth,
      points: [],
    };

    if (typeof this.options.onBegin === 'function') {
      this.options.onBegin(event);
    }

    this.data.push(newPointGroup);
    this.reset();
    this._strokeUpdate(event);
  }

  private _strokeUpdate(
    event: MouseEvent | Touch | WechatMiniprogram.Touch,
  ): void {
    if (this.data.length === 0) {
      // This can happen if clear() was called while a signature is still in progress,
      // or if there is a race condition between start/update events.
      this.strokeBegin(event);
      return;
    }

    // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
    // @ts-ignore wechatMiniProgram declaration error
    const x = event.clientX || event.x;
    // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
    // @ts-ignore wechatMiniProgram declaration error
    const y = event.clientY || event.y;

    const point = this._createPoint(x, y);
    const lastPointGroup = this.data[this.data.length - 1];
    const lastPoints = lastPointGroup.points;
    const lastPoint =
      lastPoints.length > 0 && lastPoints[lastPoints.length - 1];
    const isLastPointTooClose = lastPoint
      ? point.distanceTo(lastPoint) <= this.minDistance
      : false;
    const color = lastPointGroup.color;

    // Skip this point if it's too close to the previous one
    if (!lastPoint || !(lastPoint && isLastPointTooClose)) {
      const curve = this._addPoint(point);

      if (!lastPoint) {
        this._drawDot({ color, point });
      } else if (curve) {
        this._drawCurve({ color, curve });
      }

      lastPoints.push({
        time: point.time,
        x: point.x,
        y: point.y,
      });
      console.log("%c >>>>>>开始回掉轨迹数据","color:blue",)
      this.sendMsg && this.sendMsg(lastPointGroup);
    }
  }

  public strokeEnd(event: MouseEvent | Touch | WechatMiniprogram.Touch): void {
    this._strokeUpdate(event);
    if (typeof this.options.onEnd === 'function') {
      this.options.onEnd(event);
    }
  }

  // Called when a new line is started
  public reset(): void {
    this.lastPoints = [];
    this.lastVelocity = 0;
    this.lastWidth = (this.minWidth + this.maxWidth) / 2;
    this.ctx.fillStyle = this.penColor;
  }

  private _fromData(
    pointGroups: PointGroup[],
    drawCurve: SignaturePadBase['_drawCurve'],
    drawDot: SignaturePadBase['_drawDot'],
  ): void {
    for (const group of pointGroups) {
      const { color, points, minWidth } = group;

      if (points.length > 1) {
        for (let j = 0; j < points.length; j += 1) {
          const basicPoint = points[j];
          const point = new Point(basicPoint.x, basicPoint.y, basicPoint.time);

          // All points in the group have the same color, so it's enough to set
          // penColor just at the beginning.
          this.penColor = color;
          if (minWidth) {
            this.minWidth = minWidth;
            this.maxWidth = minWidth + 2;
          }

          if (j === 0) {
            this.reset();
          }

          const curve = this._addPoint(point);

          if (curve) {
            drawCurve({ color, curve });
          }
        }
      } else {
        this.reset();

        drawDot({
          color,
          point: points[0],
        });
      }
    }
  }

  private _toSVG(): string {
    const pointGroups = this.data;
    const ratio = Math.max(window.devicePixelRatio || 1, 1);
    const minX = 0;
    const minY = 0;
    const maxX = this.canvas.width / ratio;
    const maxY = this.canvas.height / ratio;
    const svg = document.createElementNS('http://www.w3.org/2000/svg', 'svg');

    svg.setAttribute('width', this.canvas.width.toString());
    svg.setAttribute('height', this.canvas.height.toString());

    this._fromData(
      pointGroups,

      ({ color, curve }: { color: string; curve: Bezier }) => {
        const path = document.createElement('path');

        // Need to check curve for NaN values, these pop up when drawing
        // lines on the canvas that are not continuous. E.g. Sharp corners
        // or stopping mid-stroke and than continuing without lifting mouse.
        /* eslint-disable no-restricted-globals */
        if (
          !isNaN(curve.control1.x) &&
          !isNaN(curve.control1.y) &&
          !isNaN(curve.control2.x) &&
          !isNaN(curve.control2.y)
        ) {
          const attr =
            `M ${curve.startPoint.x.toFixed(3)},${curve.startPoint.y.toFixed(
              3,
            )} ` +
            `C ${curve.control1.x.toFixed(3)},${curve.control1.y.toFixed(3)} ` +
            `${curve.control2.x.toFixed(3)},${curve.control2.y.toFixed(3)} ` +
            `${curve.endPoint.x.toFixed(3)},${curve.endPoint.y.toFixed(3)}`;
          path.setAttribute('d', attr);
          path.setAttribute('stroke-width', (curve.endWidth * 2.25).toFixed(3));
          path.setAttribute('stroke', color);
          path.setAttribute('fill', 'none');
          path.setAttribute('stroke-linecap', 'round');

          svg.appendChild(path);
        }
        /* eslint-enable no-restricted-globals */
      },

      ({ color, point }: { color: string; point: BasicPoint }) => {
        const circle = document.createElement('circle');
        const dotSize =
          typeof this.dotSize === 'function' ? this.dotSize() : this.dotSize;
        circle.setAttribute('r', dotSize.toString());
        circle.setAttribute('cx', point.x.toString());
        circle.setAttribute('cy', point.y.toString());
        circle.setAttribute('fill', color);

        svg.appendChild(circle);
      },
    );

    const prefix = 'data:image/svg+xml;base64,';
    const header =
      '<svg' +
      ' xmlns="http://www.w3.org/2000/svg"' +
      ' xmlns:xlink="http://www.w3.org/1999/xlink"' +
      ` viewBox="${minX} ${minY} ${maxX} ${maxY}"` +
      ` width="${maxX}"` +
      ` height="${maxY}"` +
      '>';
    let body = svg.innerHTML;

    // IE hack for missing innerHTML property on SVGElement
    if (body === undefined) {
      const dummy = document.createElement('dummy');
      const nodes = svg.childNodes;
      dummy.innerHTML = '';

      // tslint:disable-next-line: prefer-for-of
      for (let i = 0; i < nodes.length; i += 1) {
        dummy.appendChild(nodes[i].cloneNode(true));
      }

      body = dummy.innerHTML;
    }

    const footer = '</svg>';
    const data = header + body + footer;

    return prefix + btoa(data);
  }

  private _createPoint(x: number, y: number): Point {
    let rect;
    if (
      window &&
      window.HTMLCanvasElement &&
      this.canvas instanceof window.HTMLCanvasElement
    ) {
      rect = this.canvas.getBoundingClientRect();
    } else {
      rect = { left: 0, top: 0 };
    }

    return new Point(x - rect.left, y - rect.top, new Date().getTime());
  }

  // Add point to lastPoints array and generate a new curve if there are enough points (i.e. 3)
  protected _addPoint(point: Point): Bezier | null {
    const { lastPoints } = this;

    lastPoints.push(point);

    if (lastPoints.length > 2) {
      // To reduce the initial lag make it work with 3 points
      // by copying the first point to the beginning.
      if (lastPoints.length === 3) {
        lastPoints.unshift(lastPoints[0]);
      }

      // _points array will always have 4 points here.
      const widths = this._calculateCurveWidths(lastPoints[1], lastPoints[2]);
      const curve = Bezier.fromPoints(lastPoints, widths);

      // Remove the first element from the list, so that there are no more than 4 points at any time.
      lastPoints.shift();

      return curve;
    }

    return null;
  }

  private _calculateCurveWidths(
    startPoint: Point,
    endPoint: Point,
  ): { start: number; end: number } {
    const velocity =
      this.velocityFilterWeight * endPoint.velocityFrom(startPoint) +
      (1 - this.velocityFilterWeight) * this.lastVelocity;

    const newWidth = this._strokeWidth(velocity);

    const widths = {
      end: newWidth,
      start: this.lastWidth,
    };

    this.lastVelocity = velocity;
    this.lastWidth = newWidth;

    return widths;
  }

  private _strokeWidth(velocity: number): number {
    return Math.max(this.maxWidth / (velocity + 1), this.minWidth);
  }

  private _drawCurveSegment(x: number, y: number, width: number): void {
    const ctx = this.ctx;

    ctx.moveTo(x, y);
    ctx.arc(x, y, width, 0, 2 * Math.PI, false);
    this._isEmpty = false;
  }

  protected _drawCurve({
    color,
    curve,
  }: {
    color: string;
    curve: Bezier;
  }): void {
    const ctx = this.ctx;
    const widthDelta = curve.endWidth - curve.startWidth;
    // '2' is just an arbitrary number here. If only lenght is used, then
    // there are gaps between curve segments :/
    const drawSteps = Math.floor(curve.length()) * 2;

    ctx.beginPath();
    ctx.fillStyle = color;

    for (let i = 0; i < drawSteps; i += 1) {
      // Calculate the Bezier (x, y) coordinate for this step.
      const t = i / drawSteps;
      const tt = t * t;
      const ttt = tt * t;
      const u = 1 - t;
      const uu = u * u;
      const uuu = uu * u;

      let x = uuu * curve.startPoint.x;
      x += 3 * uu * t * curve.control1.x;
      x += 3 * u * tt * curve.control2.x;
      x += ttt * curve.endPoint.x;

      let y = uuu * curve.startPoint.y;
      y += 3 * uu * t * curve.control1.y;
      y += 3 * u * tt * curve.control2.y;
      y += ttt * curve.endPoint.y;

      const width = Math.min(
        curve.startWidth + ttt * widthDelta,
        this.maxWidth,
      );
      this._drawCurveSegment(x, y, width);
    }

    ctx.closePath();
    ctx.fill();
  }

  protected _drawDot({
    color,
    point,
  }: {
    color: string;
    point: BasicPoint;
  }): void {
    const ctx = this.ctx;
    const width =
      typeof this.dotSize === 'function' ? this.dotSize() : this.dotSize;

    ctx.beginPath();
    this._drawCurveSegment(point.x, point.y, width);
    ctx.closePath();
    ctx.fillStyle = color;
    ctx.fill();
  }
}
